ephemeral mode check

pohly opened the pull request:
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:

This makes it safer to use ephemeral inline volumes (an alpha feature in 1.15) because it prevents accidental usage of a normal CSI driver for an ephemeral inline volume. The whitelisting of CSI drivers via a pod security policy would have the same effect, but might not be active in a cluster.

**Special notes for your reviewer**:

**Which issue(s) this PR fixes**:
Fixes: https://github.com/kubernetes/kubernetes/issues/79624
Related-to: https://github.com/kubernetes/kubernetes/pull/79983, which has been merged

**Does this PR introduce a user-facing change?**:
```release-note
a CSI driver that supports ephemeral inline volumes must explicitly declare that by providing a CSIDriver object where the new "mode" field is set to "ephemeral" or "persistent+ephemeral"
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190122-csi-inline-volumes.md#driver-mode
```

pohly committed:
storage: introduce CSIDriver.Spec.VolumeLifecycleModes

Using a "normal" CSI driver for an inline ephemeral volume may have
unexpected and potentially harmful effects when the driver gets a
NodePublishVolume call that it isn't expecting. To prevent that mistake,
driver deployments for a driver that supports such volumes must:
- deploy a CSIDriver object for the driver
- list "ephemeral" as one of the supported modes

The default is "persistent", so existing deployments continue to work
and are automatically protected against incorrect usage.

This commit contains the API change. Generated code and manual code
which uses the new API follow.

pohly committed:
storage: check CSIDriver.Spec.VolumeLifecycleModes

Using a "normal" CSI driver for an inline ephemeral volume may have
unexpected and potentially harmful effects when the driver gets a
NodePublishVolume call that it isn't expecting. To prevent that mistake,
driver deployments for a driver that supports such volumes must:
- deploy a CSIDriver object for the driver
- set CSIDriver.Spec.VolumeLifecycleModes such that it contains "ephemeral"

The default for that field is "persistent", so existing deployments
continue to work and are automatically protected against incorrect
usage.

For the E2E tests we need a way to specify the driver mode. The
existing cluster-driver-registrar doesn't support that and also was
deprecated, so we stop using it altogether and instead deploy and
patch a CSIDriver object.

pohly committed:
storage: generated files for CSIDriver.Spec.VolumeLifecycleModes

pohly committed:
bazel update

pohly commented:
/priority important-soon
/cc @vladimirvivien

cwdsuzhou commented on pkg/volume/csi/csi_plugin_test.go:
```diff
@@ -652,23 +663,29 @@ func TestPluginNewMounterWithInline(t *testing.T) {
+		fakeClient := fakeclient.NewSimpleClientset(driver)
+		plug, tmpDir := newTestPlugin(t, fakeClient)
 		defer os.RemoveAll(tmpDir)
 
```
since we use defer in foreach，using t.Run is better

pohly replied:
Right, good catch. Other tests already did that and I missed that this one doesn't. Will fix it.

pohly replied:
Actually there was a `t.Run`, I just moved the code into the wrong spot. Fixed.

pohly commented:
Before actually asking for an API review, I need to rebase and split the commit into smaller ones (just the API change, generated files, additional code and tests).

Right now I am just trying to find out what I might have missed (like updating some more generated files) and why the roundtrip test is failing.

vladimirvivien commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -785,13 +798,61 @@ func (p *csiPlugin) skipAttach(driver string) (bool, error) {
+// supportsVolumeMode looks up CSIDriver object associated with the
+// driver name to determine if the driver supports the given volume
+// mode. An error indicates that it isn't supported and explains why.
+func (p *csiPlugin) supportsVolumeMode(driver string, volumeMode csiVolumeMode) error {
```
This probably would be better as a `isVolumeSuppoerted(driver, mode) (bool, error)` Returns `true` if supported, `false` if not, and `error` if something went wrong while attempting to check.

pohly replied:
I had that initially, but then there's no way to convey to the caller what the reason is when the mode isn't supported.

These two error message would have to be removed:
https://github.com/kubernetes/kubernetes/blob/1a14c40a538a6324f867420897d11b4378585941/pkg/volume/csi/csi_plugin.go#L812-L817

vladimirvivien commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -785,13 +798,61 @@ func (p *csiPlugin) skipAttach(driver string) (bool, error) {
+}
+
+// lookupDriverMode looks up CSIDriver object associated with the driver name and
+// returns what it finds there or the "persistent" default.
```
Please clarify in comment what the `bool`  return param is for is for.

pohly replied:
Will do. It tells the caller whether the returned string was explicitly set or is the default value. That's relevant when it comes to producing a good error message.

Would it be better to return the empty string for "not set" and drop the bool?

pohly replied:
> Would it be better to return the empty string for "not set" and drop the bool?

I changed it to that. @vladimirvivien: okay?

vladimirvivien commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -785,13 +798,61 @@ func (p *csiPlugin) skipAttach(driver string) (bool, error) {
+	if csiDriver.Spec.Mode == nil {
+		return storage.PersistentDriverMode, true, nil
+	}
+	return *csiDriver.Spec.Mode, false, nil
```
Should `true` be returned here since driver mode was set ?

vladimirvivien commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -785,13 +798,61 @@ func (p *csiPlugin) skipAttach(driver string) (bool, error) {
+		return "", false, err
+	}
+	if csiDriver.Spec.Mode == nil {
+		return storage.PersistentDriverMode, true, nil
```
Should `false` be returned since driver mode is not set.  the boolean needs to be clarified I suppose.

vladimirvivien reviewed (commented):
Looking good, left comments for some minor issues.

msau42 commented:
/assign @msau42 @jsafrane

msau42 commented on pkg/apis/storage/types.go:
```diff
@@ -293,8 +293,34 @@ type CSIDriverSpec struct {
+	// - "persistent+ephemeral": both modes supported
+	// More modes may be added in the future.
+	// +optional
+	Mode *CSIDriverMode
```
Validation for this new field needs to be added in https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/storage/validation/validation.go

pohly replied:
Agreed, will add this. I just wanted to fix the roundtrip error first before working on making the API nicer to use.

pohly replied:
Done.

msau42 commented on pkg/apis/storage/v1beta1/defaults.go:
```diff
@@ -47,4 +47,8 @@ func SetDefaults_CSIDriver(obj *storagev1beta1.CSIDriver) {
 		obj.Spec.PodInfoOnMount = new(bool)
 		*(obj.Spec.PodInfoOnMount) = false
 	}
+	if obj.Spec.Mode == nil {
```
I think this needs to be feature gated

msau42 replied:
We also need to drop the alpha field in `PrepareForCreate`, `PrepareForUpdate` calls in https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/storage/csidriver/strategy.go.  Take a look at [this example](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/storage/csinode/strategy.go)

pohly replied:
Okay. But will this help with the roundtrip problem?

pohly replied:
And which part of that example are the hand-written files? They all come from https://github.com/kubernetes/kubernetes/commit/bb45b8ee34f11bf858a29c8e4176b93ed1b76928#diff-da9973f9d53ca756ac31a053eb90161c, which includes everything in a single commit.

msau42 replied:
the entire strategy.go is a handwritten file

msau42 replied:
To fix the roundtripping test, I think we need to modify the [fuzzer](https://github.com/kubernetes/kubernetes/blob/6c31101257bfcd47fa53702cea07fe2eedf2ad92/pkg/apis/storage/fuzzer/fuzzer.go)

pohly replied:
I added the feature gate check to `defaults.go` and `strategy.go`. With the feature disabled, the roundtrip test passes, but of course breaks again when enabling it by default.

> To fix the roundtripping test, I think we need to modify the fuzzer

Yes, that did the trick.

I'm going to squash the following change into the commit:
```
diff --git a/pkg/apis/storage/fuzzer/fuzzer.go b/pkg/apis/storage/fuzzer/fuzzer.go
index 2081ffe846..c205f2c4f7 100644
--- a/pkg/apis/storage/fuzzer/fuzzer.go
+++ b/pkg/apis/storage/fuzzer/fuzzer.go
@@ -46,6 +46,10 @@ var Funcs = func(codecs runtimeserializer.CodecFactory) []interface{} {
                                obj.Spec.PodInfoOnMount = new(bool)
                                *(obj.Spec.PodInfoOnMount) = false
                        }
+                       if obj.Spec.Mode == nil {
+                               obj.Spec.Mode = new(storage.CSIDriverMode)
+                               *obj.Spec.Mode = storage.PersistentDriverMode
+                       }
                },
        }
 }
diff --git a/pkg/apis/storage/v1beta1/defaults.go b/pkg/apis/storage/v1beta1/defaults.go
index 05bc6dbb3e..75acec10f3 100644
--- a/pkg/apis/storage/v1beta1/defaults.go
+++ b/pkg/apis/storage/v1beta1/defaults.go
@@ -20,6 +20,8 @@ import (
        "k8s.io/api/core/v1"
        storagev1beta1 "k8s.io/api/storage/v1beta1"
        "k8s.io/apimachinery/pkg/runtime"
+       utilfeature "k8s.io/apiserver/pkg/util/feature"
+       "k8s.io/kubernetes/pkg/features"
 )
 
 func addDefaultingFuncs(scheme *runtime.Scheme) error {
@@ -47,7 +49,7 @@ func SetDefaults_CSIDriver(obj *storagev1beta1.CSIDriver) {
                obj.Spec.PodInfoOnMount = new(bool)
                *(obj.Spec.PodInfoOnMount) = false
        }
-       if obj.Spec.Mode == nil {
+       if obj.Spec.Mode == nil && utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
                obj.Spec.Mode = new(storagev1beta1.CSIDriverMode)
                *(obj.Spec.Mode) = storagev1beta1.PersistentDriverMode
        }
diff --git a/pkg/registry/storage/csidriver/strategy.go b/pkg/registry/storage/csidriver/strategy.go
index 1f534b2e76..1b31a7b23a 100644
--- a/pkg/registry/storage/csidriver/strategy.go
+++ b/pkg/registry/storage/csidriver/strategy.go
@@ -22,9 +22,11 @@ import (
        "k8s.io/apimachinery/pkg/runtime"
        "k8s.io/apimachinery/pkg/util/validation/field"
        "k8s.io/apiserver/pkg/storage/names"
+       utilfeature "k8s.io/apiserver/pkg/util/feature"
        "k8s.io/kubernetes/pkg/api/legacyscheme"
        "k8s.io/kubernetes/pkg/apis/storage"
        "k8s.io/kubernetes/pkg/apis/storage/validation"
+       "k8s.io/kubernetes/pkg/features"
 )
 
 // csiDriverStrategy implements behavior for CSIDriver objects
@@ -41,8 +43,12 @@ func (csiDriverStrategy) NamespaceScoped() bool {
        return false
 }
 
-// ResetBeforeCreate clears the Status field which is not allowed to be set by end users on creation.
+// PrepareForCreate clears the Mode field if the corresponding feature is disabled.
 func (csiDriverStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
+       if !utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
+               csiDriver := obj.(*storage.CSIDriver)
+               csiDriver.Spec.Mode = nil
+       }
 }
 
 func (csiDriverStrategy) Validate(ctx context.Context, obj runtime.Object) field.ErrorList {
@@ -62,8 +68,12 @@ func (csiDriverStrategy) AllowCreateOnUpdate() bool {
        return false
 }
 
-// PrepareForUpdate sets the Status fields which is not allowed to be set by an end user updating a CSIDriver
+// PrepareForUpdate clears the Mode field if the corresponding feature is disabled.
 func (csiDriverStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+       if !utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
+               newCSIDriver := obj.(*storage.CSIDriver)
+               newCSIDriver.Spec.Mode = nil
+       }
 }
 
 func (csiDriverStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
```

pohly committed:
storage: validate CSIDriver.Spec.VolumeLifecycleModes

This ensures that users get a good error message early on when trying
to do something that isn't okay:

  $ kubectl create -f csi-hostpath-driverinfo.yaml
  The CSIDriver "hostpath.csi.k8s.io" is invalid: spec.volumeLifecycleModes: Unsupported value: "foobar": supported values: "persistent", "ephemeral"

pohly commented:
I think I have addressed all comments and all tests are passing. I also verified manually that the round-trip test still passes when enabling the CSIInlineVolume feature. Thanks everyone for your help!

-> no longer work-in-progress, please review.

msau42 commented:
/label api-review

liggitt commented:
/assign @smarterclayton

smarterclayton commented on pkg/apis/storage/types.go:
```diff
@@ -293,8 +293,34 @@ type CSIDriverSpec struct {
+const (
+	// PersistentDriverMode explicitly confirms that the driver implements
+	// the full CSI spec. It is the default when CSIDriverSpec.Mode is not
+	// set.
```
Add some additional details here about how the naming of the mode relates to other "persistent" concepts in Kube - i.e. from discussion in api review "persistent" CSI driver modes are intended to relate to a "persistent volume claims" while "ephemeral" inline volumes do not.  I would also suggest adding a few examples of what sorts of drivers would use this mode (so that an author / casual browser can orient themselves).

so for example for ephemeral "a CSI driver that created/downloaded a vault secret and put it into a directory for a given pod might use ephemeral driver mode because they want that secret to be tied to the pod, and ephemeral gives them the ability to ...".  And then you might want to also mention that the driver is free to have a broader scope than the life of the pod, but that might be less common - "for example, an ephemeral inline CSI volume might share a temporary, garbage-collected empty dir that is only accessible to pods in one namespace".

kfox1111 replied:
Are folks going to be reading source code for this? I'm thinking this might be better done as an extension to the KEP?

pohly replied:
Let's document this properly in the API and then update the KEP accordingly. Then whatever people encounter first has a good description.

msau42 replied:
The comments in the external types are used to generate api docs.  And I believe 'kubectl explain' uses this too.  So it's good to have these comments be very thorough.

kfox1111 replied:
Ah, ok. That makes sense.

smarterclayton commented on staging/src/k8s.io/api/storage/v1beta1/types.go:
```diff
@@ -291,10 +291,41 @@ type CSIDriverSpec struct {
 	// +optional
 	PodInfoOnMount *bool `json:"podInfoOnMount,omitempty" protobuf:"bytes,2,opt,name=podInfoOnMount"`
+
+	// If set, Mode defines how this CSI volume driver can be used.
```
style: use `mode` instead of `Mode` since the godoc is intended for people who are using the API via JSON, not Go code.

smarterclayton replied:
After discussion in the review call, we (@msau42 @jsafrane and @smarterclayton) agreed this should probably be an array of enumerations, so `(DriverModes|SupportDriverModes|AllowedDriverModes) []CSIDriverMode` with enumeration values `Persistent`, `Ephemeral`.

DriverModes may be the best, please think about how a user reading this would think about it.

pohly replied:
I agree, turning this into an array makes sense.

Instead using the word "mode" (which is awfully overloaded), would it make sense turn this new field in a generic "Capabilities" array? That then opens the door for adding future enumeration values that having nothing to do with the current usage (volume "modes").

pohly replied:
Yes another option is one field per capability:
- `PersistentVolumes *bool` (default true)
- `EphemeralVolumes *bool` (default false)

pohly replied:
I actually like that last option best myself.

msau42 replied:
I prefer putting these two into a separate array, because these two capabilities and the functionality that they impact are correlated very closely with each other.

kfox1111 replied:
How would an enumeration where multiple can be used work in yaml?

PersistentVolumes and EphemeralVolumes would work for me.

Or:
Mode:
 - Persistent
 - Ephemeral

Where the array entries are strings and its existence in the list is true. Maybe thats what @smarterclayton was meaning?

pohly replied:
@kfox1111: yes, that's the proposal from https://github.com/kubernetes/kubernetes/pull/80568#discussion_r309775598.

Okay, as as the flags are meant to be semantically related, then let's go with that proposal and define it as:
```
VolumeModes:
- Persistent
- Ephemeral
```

If the list is empty or nil,  "Persistent" is added as default.

I chose this naming because IMHO it makes more sense to describe it in terms of "what does the driver support" instead of "in which mode does the driver operate", in particular when the driver switches on a volume-by-volume basis.

smarterclayton commented on staging/src/k8s.io/api/storage/v1beta1/types.go:
```diff
@@ -291,10 +291,41 @@ type CSIDriverSpec struct {
+	//
+	// Other modes are:
+	// - "ephemeral": only use this driver for inline ephemeral volumes (CSIVolumeSource)
+	// - "persistent+ephemeral": both modes supported
```
Might suggest adding an example of use here (same as above on constant) because this would be extended to the user.

smarterclayton commented on staging/src/k8s.io/api/storage/v1beta1/types.go:
```diff
@@ -291,10 +291,41 @@ type CSIDriverSpec struct {
+	// PersistentDriverMode explicitly confirms that the driver implements
+	// the full CSI spec. It is the default when CSIDriverSpec.Mode is not
+	// set.
+	PersistentDriverMode CSIDriverMode = "persistent"
```
style: API constants should always be `CamelCase`, so `Persistent`, `Ephemeral`, `All`.  We generally do not use "composition constants" (`persistent+ephemeral`).

It may be more appropriate to use an array here, since if we added a third mode in the future, we don't want to have to do `AB`, `BC`, etc.  This seems like an array of supported modes, so `[Persistent,Ephemeral]`, `[Persistent]`, etc.

pohly replied:
The effect is that in the driver info yaml, the values also have to be capitalized. Is that really intended?
```
apiVersion: storage.k8s.io/v1beta1
kind: CSIDriver
metadata:
  name: hostpath.csi.k8s.io
spec:
  # Supports both modes, but needs pod info for that to determine the actual mode.
  podInfoOnMount: true
  volumeModes:
  - Persistent
  - Ephemeral
```

Or do I have to implement some kind of mapping?

pohly replied:
Looks like this is intended, see `volumeBindingMode: Immediate` (https://kubernetes.io/docs/concepts/storage/storage-classes/#the-storageclass-resource).

smarterclayton commented on staging/src/k8s.io/api/storage/v1beta1/types.go:
```diff
@@ -291,10 +291,41 @@ type CSIDriverSpec struct {
 	// "csi.storage.k8s.io/pod.namespace": pod.Namespace
 	// "csi.storage.k8s.io/pod.uid": string(pod.UID)
+	// "csi.storage.k8s.io/ephemeral": "true" iff the volume is an ephemeral inline volumed
+	//                                 defined by a CSIVolumeSource, otherwise "false"
```
Because this is new, you also have to define the behavior of absence.  So for example,

"If this annotation is not present, the CSI implementation must assume the cluster does not support ephemeral inline volume sources, and that this volume must be persistent."

smarterclayton replied:
Doesn't have to be inline to the annotation, you could move it to a separate paragraph after the annotations are listed.

pohly replied:
Added this paragraph:

"csi.storage.k8s.io/ephemeral" is a new feature in Kubernetes 1.16. It is only
required for drivers which support both modes. Other drivers can leave pod info
disabled and/or ignore this field. As Kubernetes 1.15 doesn't support this field,
drivers can only support one mode when deployed on such a cluster and the
deployment determines which mode that is, for example via a command line parameter
of the driver.

smarterclayton commented on staging/src/k8s.io/api/testdata/HEAD/storage.k8s.io.v1beta1.CSIDriver.json:
```diff
@@ -42,6 +42,7 @@
     "attachRequired": false,
-    "podInfoOnMount": true
+    "podInfoOnMount": true,
+    "mode": "Ă凗蓏Ŋ蛊ĉy"
```
Check whether you need to generate a specific fuzz type for this mode, so that you see `Ephemeral`, `Persistent` and we can verify that both valid (constants) or invalid (the example here) are properly handled.

liggitt replied:
this fuzzer just handles serialization/deserialization, not conversion/validation

liggitt replied:
https://github.com/kubernetes/kubernetes/pull/80568#discussion_r309771148 is still good to do, but that won't affect the serialization/deserialization fixtures

pohly replied:
I don't know what is meant with "generate a specific fuzz type for this mode". Is there an example that I can follow?

This file is generated. So the content may look random and I don't know exactly what it means, but I don't think it is invalid. I assumed that it is some kind of binary encoding for the CSIDriverMode type. Does that make any sense?

smarterclayton commented on pkg/apis/storage/fuzzer/fuzzer.go:
```diff
@@ -46,6 +46,10 @@ var Funcs = func(codecs runtimeserializer.CodecFactory) []interface{} {
 			}
+			if obj.Spec.Mode == nil {
+				obj.Spec.Mode = new(storage.CSIDriverMode)
+				*obj.Spec.Mode = storage.PersistentDriverMode
```
Above this, instead of letting the fuzzer run, I would suggest adding random values based on the constants, with a sampling of valid values, valid + invalid values, invalid values, and empty.

For instance:

```
switch c.Rand.Intn(3) {
case 0:
  obj.Spec.Mode = nil
case 1:
  obj.Spec.Mode = &driverModes[c.Rand.Intn(len(driverModes))]
case 2:
  obj.Spec.Mode = CSIDriverMode(c.Rand.String())
}
```

smarterclayton commented on pkg/apis/storage/validation/validation.go:
```diff
@@ -445,3 +446,24 @@ func validatePodInfoOnMount(podInfoOnMount *bool, fldPath *field.Path) field.Err
+	allErrs := field.ErrorList{}
+	if mode != nil {
+		switch *mode {
+		case storage.PersistentDriverMode:
```
nit: golang style is to collapse these into comma delimited 

```
case storage.PersistentDriverMode, storage.EphemeralDriverMode, ...:
default:
```

since it's more canonical

smarterclayton commented on pkg/apis/storage/validation/validation_test.go:
```diff
@@ -1774,6 +1802,15 @@ func TestCSIDriverValidation(t *testing.T) {
 				PodInfoOnMount: nil,
 			},
 		},
+		{
```
also test nil, even though we default it, just in case defaulting changes in the future

pohly commented:
@smarterclayton: I'll update the PR tomorrow. Do you want me to add changes on top of it in separate commits for review and then later squash before merging, or shall I squash directly?

I also need to rebase.

msau42 commented:
I'm fine with squashing directly

msau42 commented:
recording of api review here: https://youtu.be/7xmg7uDsqfo

pohly commented:
/retest

pohly commented:
@smarterclayton: pushed an update. Please have another look.  If you want to diff against the previous revision, it's here: https://github.com/pohly/kubernetes/tree/ephemeral-mode-check-api-review-1

I asked about the `API rule violation: list_type_missing` on the api-review Slack channel. The results were inconclusive so far, therefore I settled on just whitelisting that new violation.

msau42 commented on pkg/registry/storage/csidriver/strategy.go:
```diff
@@ -62,8 +68,12 @@ func (csiDriverStrategy) AllowCreateOnUpdate() bool {
-// PrepareForUpdate sets the Status fields which is not allowed to be set by an end user updating a CSIDriver
+// PrepareForUpdate clears the Mode field if the corresponding feature is disabled.
 func (csiDriverStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
```
It should only clear it if it's not already set.  This allows the field to remain when downgrading to a version that has the feature disabled.

msau42 commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -392,10 +381,11 @@ func (p *csiPlugin) NewMounter(
 	}
 
-	// TODO(pohly): check CSIDriver.Spec.Mode to ensure that the CSI driver
+	// Check CSIDriver.Spec.Mode to ensure that the CSI driver
```
should this check be feature gated on CSIInlineVolume, since the volume mode field is controlled by that feature gate?

pohly replied:
That won't change much because the result "should" be the same also when the feature is disabled. But to avoid potential mistakes, I short-circuited the logic in `supportsVolumeMode`:
```
func (p *csiPlugin) supportsVolumeMode(driver string, volumeMode storage.CSIVolumeMode) error {
	if !utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
		// Nothing to check, feature disabled.
		return nil
	}
```

msau42 commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -801,11 +791,67 @@ func (p *csiPlugin) skipAttach(driver string) (bool, error) {
+	case containsVolumeMode(supportedModes, volumeMode):
+		// Explicitly listed.
+		return nil
+	case volumeMode == storage.PersistentVolumeMode && len(supportedModes) == 0:
```
If we gate this entire function on CSIInlineVolume, then we don't need this case because the field will be defaulted when we read the CSIDriver object.  This way we don't encode our own defaulting logic outside of API defaulting.

pohly replied:
No, not quite. This case is also hit if:
- the feature is enabled
- a driver doesn't have a CSIDriver object
- the volume mode is "persistent"

In that case it's okay to proceed with the volume. If we removed this case, we would continue with the error cases.

msau42 replied:
There's a difference between "CSIDriver object doesn't exist" and "CSIDriver does exist but is missing the field".  When the feature is enabled, the CSIDriver object should also have volumeModes set, either due to defaulting or the user explicitly setting the value.  If the field is not set, then that perhaps indicates that the feature is disabled on the apiserver, which is a configuration error.  I think in that case, we want to return an error here instead of assuming persistent mode

pohly replied:
Okay. To implement this distinction, I merged `lookupVolumeModes` into `supportsVolumeMode` because otherwise that function wouldn't have enough information. It's less code this way, too.

In several test cases for persistent volumes the missing field had to be added, because defaulting is not active when we fake the API server. I also added some more test cases for "missing mode" or "wrong mode for persistent volume".

msau42 commented on staging/src/k8s.io/api/storage/v1beta1/types.go:
```diff
@@ -291,10 +291,56 @@ type CSIDriverSpec struct {
+	//                                 defined by a CSIVolumeSource, otherwise "false"
+	//
+	// "csi.storage.k8s.io/ephemeral" is a new feature in Kubernetes 1.16. It is only
+	// required for drivers which support both modes. Other drivers can leave pod info
```
both modes => both "Persistent" and "Ephemeral" CSIVolumeModes.

msau42 replied:
also make sure comments in both types.go files are in sync

pohly replied:
Done.

msau42 commented on pkg/apis/storage/types.go:
```diff
@@ -293,8 +293,44 @@ type CSIDriverSpec struct {
 	PodInfoOnMount *bool
+
+	// VolumeModes defines what kind of volumes this CSI volume driver supports.
+	// The default if the list is empty is "Persistent", which is the usage
```
The default mode if the list is empty is "Persistent"

pohly replied:
Yep, copied from the other types.go.

vladimirvivien reviewed (commented):
Taking shape nicely. Left a few minor comments.

pohly commented:
@smarterclayton: pushed an update. Please have another look.  If you want to diff against the previous revision, it's here: https://github.com/pohly/kubernetes/tree/ephemeral-mode-check-api-review-1

I asked about the `API rule violation: list_type_missing` on the api-review Slack channel. The results were inconclusive so far, therefore I settled on just whitelisting that new violation.

pohly commented on pkg/apis/storage/validation/validation.go:
```diff
@@ -445,3 +446,21 @@ func validatePodInfoOnMount(podInfoOnMount *bool, fldPath *field.Path) field.Err
+	allErrs := field.ErrorList{}
+	for _, mode := range modes {
+		switch mode {
+		case storage.PersistentVolumeMode, storage.EphemeralVolumeMode:
```
Yes.

The driver then has to use podinfo to determine what it needs to do in `NodePublishVolume`.

pohly commented on pkg/apis/storage/validation/validation_test.go:
```diff
@@ -1736,6 +1736,49 @@ func TestCSIDriverValidation(t *testing.T) {
 				PodInfoOnMount: &notPodInfoOnMount,
 			},
 		},
+		{
```
Specifying the mode is optional, therefore such a CSIDriverInfo doesn't get rejected (I assume that is what you mean with "unhappy"). The default is "Persistent".

This is covered by the various existing cases that don't have `VolumeModes`: they remain valid.

pohly commented on pkg/apis/storage/types.go:
```diff
@@ -293,8 +293,44 @@ type CSIDriverSpec struct {
+type CSIVolumeMode string
+
+const (
+	// PersistentDriverMode explicitly confirms that the driver implements
```
Darn, I almost got all of them, but not quite. Thanks, fixed.

msau42 commented on api/api-rules/violation_exceptions.list:
```diff
@@ -344,6 +344,7 @@ API rule violation: list_type_missing,k8s.io/api/storage/v1,StorageClassList,Ite
 API rule violation: list_type_missing,k8s.io/api/storage/v1,VolumeAttachmentList,Items
 API rule violation: list_type_missing,k8s.io/api/storage/v1alpha1,VolumeAttachmentList,Items
 API rule violation: list_type_missing,k8s.io/api/storage/v1beta1,CSIDriverList,Items
+API rule violation: list_type_missing,k8s.io/api/storage/v1beta1,CSIDriverSpec,VolumeModes
```
This was discussed briefly in api-reviews slack channel. Is the conclusion that adding to the exceptions list is expected?
@apelisse @liggitt

liggitt replied:
Yes

pohly commented on pkg/apis/storage/types.go:
```diff
@@ -293,8 +293,44 @@ type CSIDriverSpec struct {
+	// defined by the CSI specification and implemented in Kubernetes via the usual
+	// PV/PVC mechanism.
+	// The other mode is "Ephemeral". In this mode, volumes are defined inline
+	// inside the pod spec with CSIVolumeSource. A driver has to be aware of this
```
Done.

pohly commented on pkg/apis/storage/types.go:
```diff
@@ -293,8 +293,44 @@ type CSIDriverSpec struct {
+	// The other mode is "Ephemeral". In this mode, volumes are defined inline
+	// inside the pod spec with CSIVolumeSource. A driver has to be aware of this
+	// because it is only going to get a NodePublishVolume call for such a volume.
+	// A driver can support one of these modes or both. More modes may be added
```
This comment lists all modes, so one can assume that there are only two. But "A driver can support one or more of these modes" sounds better, so I've changed it to that and it won't have to be changed if something else gets added :smile:

pohly commented on pkg/apis/storage/types.go:
```diff
@@ -293,8 +293,44 @@ type CSIDriverSpec struct {
+	VolumeModes []CSIVolumeMode
 }
 
+// CSIVolumeMode is an enumeration of possible usage modes for a volume
```
"CSIVolumeMode specifies how a CSI volume is used in Kubernetes."

msau42 commented on pkg/registry/storage/csidriver/strategy.go:
```diff
@@ -41,8 +43,12 @@ func (csiDriverStrategy) NamespaceScoped() bool {
-// ResetBeforeCreate clears the Status field which is not allowed to be set by end users on creation.
+// PrepareForCreate clears the VolumeModes field if the corresponding feature is disabled.
 func (csiDriverStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
```
Also update the unit tests for these changes

pohly replied:
Done.

msau42 commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -796,11 +786,72 @@ func (p *csiPlugin) skipAttach(driver string) (bool, error) {
+func (p *csiPlugin) supportsVolumeMode(driver string, volumeMode storage.CSIVolumeMode) error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
+		// Nothing to check, feature disabled.
+		return nil
```
Should this actually return success only when volumeMode is Persistent?  If a user wants ephemeral mode, but then the cluster gets downgraded to a version where the feature is disabled, we probably don't want to continue mounting the volume as persistent.

pohly replied:
You are right. So I'll revert to the version I had before: `supportsVolumeMode` always has to be called, and depending on the current volume mode and the feature flag it'll decide whether kubelet is meant to go ahead.

pohly replied:
`supportsVolumeMode` rewritten, please have another look.

msau42 commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -288,14 +288,18 @@ func (m *mockCSIDriver) PrepareTest(f *framework.Framework) (*testsuites.PerTest
+		NodeName:                 config.ClientNodeName,
+		PodInfo:                  m.podInfo,
+		CanAttach:                &m.attachable,
+		VolumeModes: []storagev1beta1.CSIVolumeMode{
```
Should the field name be DriverMode?  VolumeMode is generally used to describe filesystem vs raw block

pohly replied:
There is no such concept as a "driver mode" anymore. That was a limitation of the alpha implementation in 1.15.

We've tried to to come up with a good name, but so far the best we came up with is "[CSI]VolumeModes". I left out the "CSI" here for the sake of brevity. In the definition here, the CSIVolumeMode type clarifies which "mode" is meant, and in yaml it also reads well:

```
VolumeModes:
- Persistent
- Ephemeral
```

pohly commented:
@msau42 I just force-pushed the update.

pohly commented:
... and now also re-generated files.

pohly commented:
Force-pushed a fix for the false positive in https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/80568/pull-kubernetes-verify/1158823312860647424

msau42 commented on pkg/registry/storage/csidriver/strategy_test.go:
```diff
@@ -74,6 +78,222 @@ func TestCSIDriverStrategy(t *testing.T) {
+		result        []storage.CSIVolumeMode
+	}{
+		{
+			name:       "add mode",
```
To make sure we cover all the scenarios, it may be easier to structure the tests like:

- before: no mode, update: no mode
- before: no mode, update: add mode
- before: mode, update: change mode
- before: mode, update: drop mode

And run each with and without feature gate enabled.

pohly replied:
Done. The "downgrade" scenario is now covered by "before: mode, update: change mode" with feature disabled, because I changed the code so that it really only tests `PrepareForUpdate`, i.e. the feature gate doesn't influence the original object.

msau42 commented on pkg/registry/storage/csidriver/strategy_test.go:
```diff
@@ -74,6 +78,222 @@ func TestCSIDriverStrategy(t *testing.T) {
 	}
 }
 
+func TestCSIDriverCreate(t *testing.T) {
```
Suffix the function names with PrepareForCreate/Update, to reflect that we're only testing those methods

pohly replied:
Done.

msau42 commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -775,11 +765,70 @@ func (p *csiPlugin) skipAttach(driver string) (bool, error) {
+// An error indicates that it isn't supported and explains why.
+func (p *csiPlugin) supportsVolumeMode(driver string, volumeMode storage.CSIVolumeMode) error {
+	inlineEnabled := utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume)
+	if volumeMode == storage.EphemeralVolumeMode && !inlineEnabled {
```
I think following the switch logic below will be simpler if we do all "feature disabled" checking up here.  So something like:
```
if !inlineEnabled {
  if volumeMode != storage.PersistentVolumeMode {
    return err
  }
  return nil
}

pohly replied:
Done.

msau42 commented on pkg/volume/csi/csi_plugin_test.go:
```diff
@@ -655,23 +685,31 @@ func TestPluginNewMounterWithInline(t *testing.T) {
 		t.Run(test.name, func(t *testing.T) {
-			plug, tmpDir := newTestPlugin(t, nil)
+			// The fake driver currently supports all modes.
+			volumeModes := []storagev1beta1.CSIVolumeMode{
```
should we make this configurable per test case so we can cover all of the scenarios in `supportsVolumeMode`?

pohly replied:
All test cases are now tested against a driver that supports both modes, only ephemeral and only persistent.

pohly commented:
@msau42 update pushed

pohly commented:
/retest

msau42 commented on pkg/apis/storage/validation/validation_test.go:
```diff
@@ -1726,6 +1726,49 @@ func TestCSIDriverValidation(t *testing.T) {
 				PodInfoOnMount: &notPodInfoOnMount,
```
I realized we're missing a unit test for `ValidateCSIDriverUpdate`. Could you look into adding one that tries to modify volumeModes?

pohly replied:
Okay.

pohly replied:
Still open, will follow shortly.

pohly replied:
Done, in a separate commit because it isn't directly related to some of the other changes.

msau42 commented on pkg/volume/csi/csi_plugin_test.go:
```diff
@@ -655,93 +695,115 @@ func TestPluginNewMounterWithInline(t *testing.T) {
-			if csiMounter.csiVolumeMode != test.csiVolumeMode {
-				t.Error("unexpected driver mode:", csiMounter.csiVolumeMode)
-			}
+				// Some test cases are broken.
```
what does this comment mean? can you clarify?

pohly replied:
"Some test cases are meant to fail because their input data is broken".

pohly replied:
Changed, will push shortly.

msau42 commented on pkg/volume/csi/csi_plugin_test.go:
```diff
@@ -655,93 +695,115 @@ func TestPluginNewMounterWithInline(t *testing.T) {
+					t.Fatal("Unexpected error:", err)
+				}
+				if shouldFail && err != nil {
+					t.Log(err)
```
Do we need to log? This may be confusing to see in the test logs since this is the success case.

pohly replied:
This isn't from me. The original code is from `TestPluginNewMounter` (
https://github.com/kubernetes/kubernetes/blob/93bc57e726f74f354b2d97a0d61f9a8310e636b7/pkg/volume/csi/csi_plugin_test.go#L596) but several other tests also do the same.

I prefer to keep it also in the new test, for the sake of consistency.

msau42 commented on test/e2e/testing-manifests/storage-csi/mock/csi-mock-driverinfo.yaml:
```diff
@@ -0,0 +1,7 @@
+apiVersion: storage.k8s.io/v1beta1
```
We can do it as a follow-on, but it would be nice to enable the test for hostpath driver too

pohly replied:
This is on my todo list. It cannot be done at the moment because we first need to merge this PR, then do a release of the hostpath driver with a driver info that specifies both modes.

pohly replied:
Let me clarify: I was thinking of the kubernetes-csi Prow jobs. For the Kubernetes E2E, testing with the hostpath driver gets enabled in ab6c7bb080467e23f5eb3dfb0c409c6316305cd3, which is part of this PR.

pohly commented on pkg/apis/storage/v1beta1/defaults.go:
```diff
@@ -47,4 +49,7 @@ func SetDefaults_CSIDriver(obj *storagev1beta1.CSIDriver) {
 		obj.Spec.PodInfoOnMount = new(bool)
 		*(obj.Spec.PodInfoOnMount) = false
 	}
+	if len(obj.Spec.VolumeModes) == 0 && utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
```
Okay.

pohly replied:
Still open, will follow shortly.

pohly commented on pkg/apis/storage/types.go:
```diff
@@ -293,8 +293,45 @@ type CSIDriverSpec struct {
+	// A driver can support one or more of these modes. More modes may be added
+	// in the future.
+	// +optional
+	VolumeModes []CSIVolumeMode
```
@msau42 So you propose `VolumeLifecycleModes []VolumeLifecycleMode`?

kfox1111 replied:
VolumeLifecycleModes works for me.

msau42 replied:
Yes this sounds good to me

pohly replied:
Name change done and pushed.

pohly commented on pkg/apis/storage/types.go:
```diff
@@ -293,8 +293,45 @@ type CSIDriverSpec struct {
+	// The other mode is "Ephemeral". In this mode, volumes are defined inline
+	// inside the pod spec with CSIVolumeSource and their lifecycle is tied to
+	// the lifecycle of that pod. A driver has to be aware of this
+	// because it is only going to get a NodePublishVolume call for such a volume.
```
I can add it, but I have no idea how self-contained the API documentation is supposed to be. We already started adding a lot of text here that duplicates information found elsewhere. If we now continued with adding links, we might as well remove the duplicate text and just say "for more information, see ...".

msau42 replied:
We are trying to strive towards having the API documentation be very descriptive so that a user reading it without much background on the subject can get a decent idea what the API is supposed to do.

So I think the comments here are good, and then for further information on how to actually implement this feature in your csi driver, we can refer to the csi driver dev docs.

pohly replied:
Done.

pohly committed:
storage: add test for ValidateCSIDriverUpdate

Adding this test was forgotten when adding CSIDriver.

pohly commented:
@msau42: I pushed another update, I think I am now done with addressing the latest round of comments.

pohly commented:
/test pull-kubernetes-e2e-gce-csi-serial
/test pull-kubernetes-integration

msau42 commented on pkg/apis/storage/validation/validation_test.go:
```diff
@@ -1772,3 +1826,94 @@ func TestCSIDriverValidation(t *testing.T) {
+	}
+
+	errorCases := []storage.CSIDriver{
+		{
```
These first few error test cases are actually testing two changes:
* Changing name or AttachedRequired, PodInfoOnMount fields
* Dropping VolumeLifecycleModes

It would be better if the error case only changed one thing at a time so that one failure reason doesn't mask the other.

msau42 commented on pkg/apis/storage/validation/validation_test.go:
```diff
@@ -1772,3 +1826,94 @@ func TestCSIDriverValidation(t *testing.T) {
+			},
+		},
+		{
+			// AttachRequired not set
```
update comment

msau42 reviewed (commented):
/lgtm
/approve

I just have two minor comments that could be addressed as followups if needed.

pohly commented:
/retest

jsafrane commented:
/retest

jsafrane commented on pkg/registry/storage/csidriver/strategy_test.go:
```diff
@@ -74,6 +78,169 @@ func TestCSIDriverStrategy(t *testing.T) {
+		runAll(t, true)
+	})
+	t.Run("without inline volumes", func(t *testing.T) {
+		runAll(t, true)
```
should be `false`?

pohly replied:
Indeed. I'll fix that in the next PR or if when there is another reason to update this PR.

pohly replied:
Fixed.

I had to revise this PR because the CSI mock driver info yaml file wasn't part of it (I incorrectly added it in a later commit that is going to be part of the future testing PR).

@msau42, @jsafrane: please re-add your lgtm.

jsafrane reviewed (commented):
/lgtm
with one minor unit test nit

pohly commented:
/retest

msau42 commented:
/lgtm
I also had some minor comments [here](https://github.com/kubernetes/kubernetes/pull/80568#discussion_r312241669) but they can be addressed separately.

pohly commented:
> I also had some minor comments here but they can be addressed separately.

Yes; those will take a bit more time than I have now.

smarterclayton commented on pkg/apis/storage/validation/validation.go:
```diff
@@ -441,3 +442,21 @@ func validatePodInfoOnMount(podInfoOnMount *bool, fldPath *field.Path) field.Err
+
+// validateVolumeLifecycleModes tests if mode has one of the allowed values.
+func validateVolumeLifecycleModes(modes []storage.VolumeLifecycleMode, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
```
Validation needs to be gated on feature flag enablement.  If the flag is off, this validation should be a no-op.

pohly replied:
When doing that, this (currently) invalid test case no longer triggers an error:
```
		{
			// invalid mode
			ObjectMeta: metav1.ObjectMeta{Name: driverName},
			Spec: storage.CSIDriverSpec{
				AttachRequired: &attachNotRequired,
				PodInfoOnMount: &notPodInfoOnMount,
				VolumeLifecycleModes: []storage.VolumeLifecycleMode{
					"no-such-mode",
				},
			},
		},
```

Is that really intended? Just wondering, I already updated `TestCSIDriverValidation` accordingly and now run it with and without the feature enabled, to cover both scenarios.

See commit a5bd16d9fba9686e0e49ec53bf7779dd83adc7cf.

I'll change `TestCSIDriverValidationUpdate` in the upcoming testing PR. In the meantime, can we merge this PR as it is now? I'm currently on vacation and would prefer to get this merged with just those changes that are considered essential.

It also continues to conflict with other changes going into master (like 7c1f55f4003ae24ffbdbb8ca6f6594807783230a). I rebased to resolved that conflict. @msau42, can you re-add your lgtm?

pohly replied:
Wait, this also affect `TestCSIDriverValidation/invalid_volume_mode` in `strategy_test.go`. I'll address that now.

pohly replied:
Done.

smarterclayton replied:
The scenario is that someone can turn the gate on, get a value stored, then turn the gate off.  If the gate is off, the object from storage must be considered valid even if the field is set, or set incorrectly.

liggitt replied:
> Validation needs to be gated on feature flag enablement. If the flag is off, this validation should be a no-op.

there are a couple existing exceptions in the codebase, but in general, validation should not pay attention to feature gates.

The recommended approach:
1. In PrepareForCreate data related to a disabled feature
2. In PrepareForUpdate data related to a disabled feature *unless* the existing object already made use of that data
3. In validation, validate normally without regard to feature gates

liggitt replied:
see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#alpha-field-in-existing-api-version for details

smarterclayton replied:
Thanks for the clarification Jordan

smarterclayton replied:
Please remove the gate as Jordan describes.

smarterclayton replied:
You may be able to drop this entire commit (if there were no new tests added)

pohly replied:
I removed the commit.

@smarterclayton, @msau42 : please check again and lgtm if it's ready.

pohly commented:
/retest

pohly commented:
/retest

msau42 commented:
/lgtm

pohly commented:
@smarterclayton: ping - is the API okay?

smarterclayton commented:
Need to remove the validation gating as Jordan described.

kfox1111 commented:
Is this the last thing needed to get ephemeral to beta?

msau42 commented:
@kfox1111 feature gate needs to be flipped + docs

pohly commented:
/retest

msau42 commented:
/lgtm

pohly commented:
@smarterclayton: sorry for the delay, I am on vacation and missed the new comments last week. The PR should now be as discussed.
